### DISCUSSION TITLE
What are the maximum number of nodes we can expect GLSP to handle in a performant way?

### DISCUSSION THREAD

**Discussion by vatsal-uppal-1997** [eclipse-glsp/glsp#904] [Q&A] [answered]
Hi All,

I am working on integrating eclipse GLSP with our Theia IDE implementation and we plan to show a large amount of graphical nodes on the screen. 

I wanted to ask that how many nodes we can expect GLSP to reasonable handle in a performant way?

We have a scenario where we have to show a thousand nodes on the screen, is this something we can expect and can we do anything from our side to improve the performance?

**Comment by planger** [comment, marked as answer]
Hi,

this is very hard to answer generically, but in general ~ thousand nodes in one diagram should be feasible. We did some evaluations in customer projects and with [Ecore](https://www.eclipse.org/glsp/gallery/#largeandzooming). Our take-aways were that rendering is the most prominent factor with respect to a snappy user experience. So we recommend to:

* Make sure your views return `undefined` if they [aren't visible](https://github.com/eclipse-sprotty/sprotty/blob/65469e49072b9a1874f4cdb1846fb25eabf2a4a8/examples/classdiagram/src/views.tsx#L28). This avoids rendering them even though they wouldn't be in the visible viewport anyway. This reduces the rendering significantly in most practical scenarios as users typically don't actually see thousands of elements even though they would work in a diagram with thousands of elements.
* If users zoom out and have the entire diagram in the viewport, you still have to render a high number of views, even though users can't actually see them in all detail. So it is recommended to reduce the level of detail by omitting to render certain child elements (e.g. icons or labels) and with that again significantly reduce the number of views to be actually rendered. You can [query the zoom factor](https://github.com/eclipse-glsp/glsp-client/blob/f5e3f3cfba45003c18bb3c74031b3e7624fc2e96/examples/workflow-glsp/src/workflow-views.tsx#L66) in the view implementations and return `undefined`, if it doesn't make sense to render a view when zoomed out.

The performance of server operations heavily depend on your specific operation and action handler implementations. So for those implementations the usual recommendations apply, such as avoiding iterating through all elements, working with indexes, etc.
